frb_codegen/src/main.rs

```diff
@@ -77,6 +70,99 @@ mod tests {
+                if e.kind() == std::io::ErrorKind::NotFound {
+                    format!(
+                        "`{test_case}`: Failed to execute 'cargo': program not found on {OS}",
+                        OS = std::env::consts::OS
+                    )
+                } else {
+                    format!(
+                        "`{test_case}`: Failed to execute 'cargo': {error} on {OS}",
+                        error = e,
+                        OS = std::env::consts::OS
+                    )
+                }
+            })
+            .unwrap()
+            .wait()
+            .map_err(|e| {
+                format!(
+                    "`{test_case}`: Failed to wait for 'cargo': {error} on {OS}",
+                    error = e,
+                    OS = std::env::consts::OS
+                )
+            })
+            .unwrap();
+        status
+    }
+
+    fn run_dart_test_command(test_case: &str, absolute_path: PathBuf) -> std::process::ExitStatus {
+        // 1.decide which dart command is valid in the specific system
+        let mut dart = if cfg!(target_os = "windows") {
+            "dart.bat"
+        } else {
+            "dart"
+        };
+        // check command validation
+        if std::process::Command::new(dart)
+            .arg("--version")
+            .status()
+            .is_err()
+        {
+            dart = if cfg!(target_os = "windows") {
+                "dart"
+            } else {
+                "dart.bat"
+            };
+
+            // check command validation again
+            if std::process::Command::new(dart)
+                .arg("--version")
+                .status()
+                .is_err()
+            {
+                panic!("Failed to find 'dart' or 'dart.bat' command in the system path.");
+            }
+        }
+
+        // 2. do the dart test check
+        let status = std::process::Command::new(dart)
+            .arg(format!("../frb_example/{test_case}/dart/lib/main.dart"))
+            .arg(absolute_path)
+            .spawn()
+            .map_err(|e| {
+                if e.kind() == std::io::ErrorKind::NotFound {
+                    format!(
+                        "`{test_case}`: Failed to execute '{dart}': program not found on {OS}",
+                        OS = std::env::consts::OS
+                    )
+                } else {
+                    format!(
+                        "`{test_case}`: Failed to execute '{dart}': {error} on {OS}",
+                        error = e,
+                        OS = std::env::consts::OS
+                    )
+                }
+            })
+            .unwrap()
+            .wait()
+            .map_err(|e| {
+                format!(
+                    "`{test_case}`: Failed to wait for '{dart}': {error} on {OS}",
+                    error = e,
+                    OS = std::env::consts::OS
+                )
+            })
+            .unwrap();
+        status
+    }
+
```

Comment:
nits:

1. maybe we can extract it to a helper file? (and only import it when in test)
2. we already have some utility functions for executing shell commands (we execute a lot, such as dart format, cbindgen, ...). is it possible we reuse them?